File: Labs/Lab4/Lab4_IterationSolution.py
# This solution handles a special case where the tree provided is too large for the recursive approach to work within memory constraints. To handle deep trees, we can employ iteration using a queue. 

from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Lab4:
    def __init__(self):
        pass

    def invertTree(self, root):
        if not root:
            return None
        queue = deque()
        queue.append(root)
        
        while queue:
            current = queue.popleft()
            current.left, current.right = current.right, current.left
            
            if current.left:
                queue.append(current.left)
            if current.right:
                queue.append(current.right)
        
        return root

File: Labs/Lab4/Lab4_IterationSolution.java
// This solution handles a special case where the tree provided is too large for the recursive approach to work within memory constraints. To handle deep trees, we can employ iteration using a queue.

import java.util.LinkedList;
import java.util.Queue;

public class Lab4 {
    public Lab4() {}

    public TreeNode invertTree(TreeNode root) {
        if (root == null) {
            return null; 
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            TreeNode current = queue.poll();
            TreeNode temp = current.left;
            current.left = current.right;
            current.right = temp;

            if (current.left != null) {
                queue.offer(current.left);
            }
            if (current.right != null) {
                queue.offer(current.right);
            }
        }
        return root;
    }
}
